internal/argocd/cluster/cluster.go

```diff
@@ -0,0 +1,78 @@
+package cluster
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/redis/go-redis/v9"
+
+	"github.com/argoproj/argo-cd/v2/common"
+	appv1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	cacheutil "github.com/argoproj/argo-cd/v2/util/cache"
+	appstatecache "github.com/argoproj/argo-cd/v2/util/cache/appstate"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	ConnectionStatusConnected    = "Connected"
+	ConnectionStatusDisconnected = "Disconnected"
+)
+
+func (m *Manager) UpdateClusterInfo(agentName, status string) {
+	// Check if we have a mapping for the requested agent
+	cluster := m.mapping(agentName)
+	if cluster == nil {
+		log().Errorf("Agent %s is not mapped to any cluster", agentName)
+		return
+	}
+
+	compression, err := cacheutil.CompressionTypeFromString(m.redisCompressionType)
+	if err != nil {
+		log().Error(err)
+		return
+	}
+
+	// Create Redis client and cache
+	redisOptions := m.setRedisAddress(agentName, cluster.Name)
+
+	if err := common.SetOptionalRedisPasswordFromKubeConfig(m.ctx, m.kubeclient, m.namespace, redisOptions); err != nil {
+		log().Errorf("Failed to fetch & set redis password for namespace %s: %v", m.namespace, err)
+	}
+	redisClient := redis.NewClient(redisOptions)
+
+	cache := appstatecache.NewCache(cacheutil.NewCache(
+		cacheutil.NewRedisCache(redisClient, time.Hour, compression)), time.Hour)
+
+	// update the info
+	if err := cache.SetClusterInfo(cluster.Server,
+		&appv1.ClusterInfo{
+			ConnectionState: appv1.ConnectionState{
+				Status:     status,
+				Message:    fmt.Sprintf("Agent: '%s' is %s with principal", agentName, strings.ToLower(status)),
+				ModifiedAt: &metav1.Time{Time: time.Now()}}},
+	); err != nil {
+		log().Errorf("Failed to update connection info in Cluster: '%s' mapped with Agent: '%s'. Error: %v", cluster.Name, agentName, err)
+		return
+	}
+
+	log().Infof("Update connection status to '%s' in Cluster: '%s' mapped with Agent: '%s'", status, cluster.Name, agentName)
+}
+
+// setRedisAddress sets Redis server address from control plane
+func (m *Manager) setRedisAddress(agentName, clusterName string) *redis.Options {
```

Comment:
Not sure if this approach to get hostname is reliable.

Reply:
I renamed the function, it sets Redis service hostname for creating a redis cache client. As of now I am using external IP to access redis server since we are running principal locally. later we can use default internal address when we have everything running in clusters. There is another option is that we set `ARGOCD_PRINCIPAL_REDIS_SERVER_ADDRESS` env variable through principal startup script.